### PR TITLE
[MM-26812] Add bot-mention and bot leave/join subscriptions

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR adds subscriptions for:

- `bot_mentioned`
- `bot_joined_channel`
- `bot_left_channel`
- `bot_joined_team`
- `bot_left_team`

For the `bot_mention` subscription, bot channel membership is checked before sending the notification.

Also added tests for the subscriptions in general

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-36812
Fixes https://mattermost.atlassian.net/browse/MM-36813

### CODE REVIEW

File: server/proxy/notify.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+func (p *Proxy) Notify(cc *apps.Context, subj apps.Subject) error {
+	subs, err := p.store.Subscription.Get(subj, cc.TeamID, cc.ChannelID)
+	if err != nil {
+		return err
+	}
+
+	return p.notify(cc, subs)
+}
+
+func (p *Proxy) NotifyRemoteWebhook(app *apps.App, data []byte, webhookPath string) error {
```
Comment (mickmister) [inline server/proxy/notify.go:24]: This function is unchanged, and is only part of this diff because it moved files.

File: server/proxy/proxy.go [hunk 2]
```diff
@@ -190,6 +195,101 @@ func (p *Proxy) NotifyRemoteWebhook(app *apps.App, data []byte, webhookPath stri
 	return upstream.Notify(up, creq)
 }
 
+func (p *Proxy) NotifyMessageHasBeenPosted(post *model.Post, cc *apps.Context) error {
```
Comment (larkox) [inline server/proxy/proxy.go:198]: I would say this function name is misleading. We are not notifying whether a message has been posted. We are notifying a bot mention.

File: server/proxy/proxy.go [hunk 3]
```diff
@@ -116,6 +117,10 @@ func (p *Proxy) Notify(cc *apps.Context, subj apps.Subject) error {
 		return err
 	}
 
+	return p.notify(cc, subs)
+}
+
+func (p *Proxy) notify(cc *apps.Context, subs []*apps.Subscription) error {
 	expander := p.newExpander(cc, p.mm, p.conf, p.store, "")
 
 	notify := func(sub *apps.Subscription) error {
```
Comment (larkox) [inline server/proxy/proxy.go:126]: This may be considered as shadowing the function. Does it make sense to use a different name?

**Review by larkox — APPROVED** [review]
LGTM, just two small nits.

File: server/proxy/proxy.go (hunk 2 above)
Comment (levb) [reply server/proxy/proxy.go:198]: I think the intention is for this function to handle the other notification as well, eventually?

File: server/proxy/proxy.go [hunk 4]
```diff
@@ -190,6 +195,101 @@ func (p *Proxy) NotifyRemoteWebhook(app *apps.App, data []byte, webhookPath stri
...
+
+		if app.BotUserID == cc.UserID {
+			subs = append(subs, sub)
+		}
+	}
+
+	return p.notify(cc, subs)
+}
+
+func (p *Proxy) canReadChannel(userID, channelID string) bool {
```
Comment (levb) [inline server/proxy/proxy.go:289]: 0/5 this function is useless, `p.mm.User.HasPermissionToChannel(userID, channelID, model.PERMISSION_READ_CHANNEL)` is perfectly intention-revealing as is

**Review by levb — APPROVED** [review]

File: server/proxy/proxy.go (hunk 2 above)
Comment (mickmister) [reply server/proxy/proxy.go:198]: @larkox @levb The function actually handles both the standard `created_post` subscription and `bot_mention` subscription. The name is meant to be a one-to-one match with the plugin hook, but prefixed with `Notify` to show that it is not the hook itself.

**Comment by mickmister** [comment]
I've added a `GET subscriptions` route to be able to examine the saved subscriptions. Need to follow up with https://mattermost.atlassian.net/browse/MM-37447 `Apps subscriptions should only be createable using the App's bot token`

**Comment by levb** [comment]
I removed myself in favor of @hanzei

File: apps/subscription.go [hunk 5]
```diff
@@ -24,13 +24,25 @@ const (
 	SubjectUserJoinedChannel Subject = "user_joined_channel"
 	SubjectUserLeftChannel   Subject = "user_left_channel"
 
+	// SubjectBotJoinedChannel and SubjectBotLeftChannel subscribes to
+	// SubjectUserJoinedChannel and SubjectUserLeftChanne plugin events,
+	// specifically when the App's bot is added or removed from the team
```
Comment (jfrerich) [inline apps/subscription.go:29]:
```suggestion
	// specifically when the App's bot is added or removed from the channel
```

File: apps/subscription.go [hunk 6]
```diff
@@ -24,13 +24,25 @@ const (
 	SubjectUserJoinedChannel Subject = "user_joined_channel"
 	SubjectUserLeftChannel   Subject = "user_left_channel"
 
+	// SubjectBotJoinedChannel and SubjectBotLeftChannel subscribes to
+	// SubjectUserJoinedChannel and SubjectUserLeftChanne plugin events,
```
Comment (jfrerich) [inline apps/subscription.go:28]:
```suggestion
	// SubjectUserJoinedChannel and SubjectUserLeftChannel plugin events,
```

**Review by jfrerich — APPROVED** [review]
Nice work, @mickmister.  I did a quick scan sense you already have 2 passing dev reviews and only found some non-code related nitpicks.

**Comment by mickmister** [comment]
@DHaussermann Please see testing notes below and leave feedback. Thank you 🙂 

Scope

Bot mention notification is sent to app when:
- the app is subscribed to the `bot_mention` subscription
- the bot is mentioned in a channel the bot is a member of

-----

Testing notes

- When app has subscribed to the bot mention event:
  - app will receive call when:
    - bot must be a member of channel
    - bot is mentioned in post
    - app only receives one call when the bot is mentioned more than once
    - when the bot and another app's bot is mentioned

  - app will not receive call when:
    - the bot is not mentioned in post
    - when another bot is mentioned, but this bot is not mentioned
- When app has not subscribed to the bot mention event:
  - app should never receive a call in any scenarios above

-----

Follow-up concerns

- Authentication for subscriptions API is currently just that the token being used must be from a sysadmin. This will change once we have OAuth scopes for the bots.
- Bot must be a member of the channel to receive the bot mention. This may not be obvious to users.

-----

Screenshots


image

image

**Comment by mickmister** [comment]
`/test subscriptions bot_mention subscribe` and `/test subscriptions bot_mention unsubscribe` commands added to the test app here: https://github.com/larkox/mattermost-app-test/pull/7

**Comment by mickmister** [comment]
@DHaussermann I've added some commands to test the join/leave functionality here https://github.com/larkox/mattermost-app-test/pull/8

**Review by DHaussermann — APPROVED** [review]

**Comment by DHaussermann** [comment]
Tested and passed
- All subscription events can be delivered
- All subscription events can be successfully enabled and disabled
- No issues found

LGTM!

**Comment by hanzei** [comment]
@mickmister Is there a docs PR for this change?

File: server/httpin/restapi/restapi.go [hunk 7]
```diff
@@ -36,6 +36,7 @@ func Init(router *mux.Router, conf config.Service, proxy proxy.Service, appServi
 		httputils.CheckAuthorized(mm, a.handleCall)).Methods("POST")
 
 	subrouter.HandleFunc(mmclient.PathSubscribe, a.handleSubscribe).Methods("POST")
+	subrouter.HandleFunc(mmclient.PathSubscribe, a.handleGetSubscriptions).Methods("GET")
```
Comment (hanzei) [inline server/httpin/restapi/restapi.go:39]: Is there a corresponding method in `ClientPP`?

File: server/proxy/proxy.go [hunk 8]
```diff
@@ -193,6 +200,100 @@ func (p *Proxy) NotifyRemoteWebhook(app *apps.App, data []byte, webhookPath stri
...
+		appsMap := p.store.App.AsMap()
+		mentionSubs, err := p.store.Subscription.Get(apps.SubjectBotMentioned, cc.TeamID, cc.ChannelID)
+		if err != nil && err != utils.ErrNotFound {
+			return errors.Wrap(err, "failed to get bot_mentioned subscriptions")
+		}
+
+		for _, sub := range mentionSubs {
+			app := appsMap[sub.AppID]
+			if app == nil {
+				continue
```
Comment (hanzei) [inline server/proxy/proxy.go:224]: Should we log an error here? I was expecting subscriptions to be cleaned when an app is uninstalled and hence this case should not happen, right?

Comment (mickmister) [reply server/proxy/proxy.go:224]: Indeed the subscriptions are not being cleaned up when an app is uninstalled. Created a ticket for this here https://mattermost.atlassian.net/browse/MM-37930

https://github.com/mattermost/mattermost-plugin-apps/blob/dfcace445d39fa39a06bd11d8f959cacadb589f0/server/proxy/uninstall.go#L15

File: server/httpin/restapi/restapi.go (hunk 7 above)
Comment (mickmister) [reply server/httpin/restapi/restapi.go:39]: @hanzei This was mainly for debugging, but it should be put there you're right. I'll follow up with this change.

FYI we will likely be reworking how subscribing works, see https://community-daily.mattermost.com/core/pl/wu83zyfgdidgxbsdc9cftpyuto and https://mattermost.atlassian.net/browse/MM-37929

**Comment by mickmister** [comment]
> @mickmister Is there a docs PR for this change?

@hanzei I'm hesitant to make docs for it because the way the App subscribes may change, but we don't have any docs on subscriptions so I'll make a docs page. Existing ticket for this here https://mattermost.atlassian.net/browse/MM-36970

**Comment by hanzei** [comment]
@mickmister Got it. Just wanted to check that we have a ticket.

**Pull request merged**